src/main/scala/org/spartanz/parserz/ParsersModule2.scala

```diff
@@ -0,0 +1,326 @@
+      private[parserz] case class Sep[SI, SO, E, A, S](value: Grammar[SI, SO, E, A], sep: Grammar[SI, SO, E, S]) extends Grammar[SI, SO, E, SeparatedBy[A, S]]
+    }
+    // format: on
+
+    final val unit: Grammar[Any, Nothing, Nothing, scala.Unit] =
+      GADT.Produce(())
+
+    final def succeed[A](a: A): Grammar[Any, Nothing, Nothing, A] =
+      GADT.Produce(a)
+
+    final def fail[E, A](e: E): Grammar[Any, Nothing, E, A] =
+      unit.mapPartial(e)(PartialFunction.empty, PartialFunction.empty)
+
+    final def consume[E, A](e: E, codec: InOut.One[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.Consume(codec.consumer, codec.condition, e)
+
+    final def consume[E, A](e: E, codec: InOut.Many[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.ConsumeMany(codec.consumer, e)
+
+    final def consumeToken[E, A](e: E, codec: InOut.ManyConditional[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.ConsumeToken(codec.consumer, e)
+
+    final def delay[SI, SO, E, A](g: => Grammar[SI, SO, E, A]): Grammar[SI, SO, E, A] =
+      GADT.Delay(() => g)
+
+    private def asEither[E, A, B](e: E)(f: A => Option[B]): A => E \/ B =
+      f(_).map(Right(_)).getOrElse(Left(e))
+  }
+
+  trait GrammarSyntax {
+
+    implicit final class ToStringOps1(self: String) {
+
+      def @@ [SI, SO, E, A](g: Grammar[SI, SO, E, A]): Grammar[SI, SO, E, A] =
+        g @@ self
+    }
+
+    implicit final class ToZipOps1[SI, SO, E, A, B](self: (Grammar[SI, SO, E, A], A)) {
+
+      def ~> [SI1 <: SI, SO1 >: SO, E1 >: E](that: Grammar[SI1, SO1, E1, B]): Grammar[SI1, SO1, E1, B] =
+        self._1.zipR(self._2, that)
+    }
+
+    implicit final class ToGrammarOps1[SI, SO, E, A, B](self: Grammar[SI, SO, E, List[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, List[A]] =
+        self.recover(Nil)
+    }
+
+    implicit final class ToGrammarOps2[SI, SO, E, A: ClassTag, B](self: Grammar[SI, SO, E, Array[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, Array[A]] =
+        self.recover(Array.empty[A])
+    }
+
+    implicit final class ToGrammarOps3[SI, SO, E, A, B](self: Grammar[SI, SO, E, (A, List[B])]) {
+
+      def foldLeft(fold: (A, B) => A, unfold: A =?> (A, B)): Grammar[SI, SO, E, A] =
+        self.map(
+          arg => {
+            arg._2.foldLeft(arg._1)(fold)
+          },
+          arg => {
+            @tailrec
+            def rec(acc: List[B])(a: A): (A, List[B]) =
+              unfold.lift(a) match {
+                case None          => (a, acc)
+                case Some((a1, b)) => rec(b :: acc)(a1)
+              }
+            rec(Nil)(arg)
+          }
+        )
+    }
+  }
+
+
+
+  private class ParserState[+E](val input: Input, var i: Int, var e: List[InOut.Error])
+
+  final def parser[S, E, A](grammar: Grammar[S, S, E, A]): Input => E \/ A =
+    input => step(grammar, new ParserState(input, 0, Nil))
+
+  private def step[S, E, A](grammar: Grammar[S, S, E, A], ps: ParserState[E]): E \/ A =
```

Comment:
Add `@tailrec` after the deletion of `Either` return value.

Reply:
I don't think it's possible due to `alt` and `zip` - I need to examine the returned values of several branches.